File: services/23-slim/src/Auth/JwtService.php
<?php

namespace App\Auth;

use Firebase\JWT\JWT;
use Firebase\JWT\Key;
use Firebase\JWT\ExpiredException;
use Firebase\JWT\SignatureInvalidException;

/**
 * JwtService — HS256 sign and verify.
 *
 * Algorithm: HS256
 * Expiry: 8 hours
 * Secret source: JWT_SECRET environment variable
 */
class JwtService
{
    private const ALGORITHM = 'HS256';
    private const TTL_SECONDS = 8 * 3600; // 8 hours

    public static function encode(array $payload): string
    {
        $now = time();
        return JWT::encode(
            array_merge($payload, ['iat' => $now, 'exp' => $now + self::TTL_SECONDS]),
            self::secret(),
            self::ALGORITHM
        );
    }

    public static function decode(string $token): ?object
    {
        try {
            return JWT::decode($token, new Key(self::secret(), self::ALGORITHM));
        } catch (ExpiredException | SignatureInvalidException | \UnexpectedValueException) {
            return null;
        }
    }

    private static function secret(): string
    {
        $secret = $_ENV['JWT_SECRET'] ?? getenv('JWT_SECRET');
        if (empty($secret)) {
            throw new \RuntimeException('JWT_SECRET environment variable is not set');
        }
        return $secret;
    }
}

File: services/22-sinatra/lib/jwt_service.rb
require 'jwt'

# JwtService — sign and verify HS256 tokens.
#
# Algorithm: HS256
# Expiry: 8 hours
# Secret source: JWT_SECRET environment variable
module JwtService
  ALGORITHM = 'HS256'
  TTL_SECONDS = 8 * 60 * 60 # 8 hours

  def self.encode(payload)
    exp = Time.now.to_i + TTL_SECONDS
    claims = payload.merge(exp: exp, iat: Time.now.to_i)
    JWT.encode(claims, secret, ALGORITHM)
  end

  def self.decode(token)
    decoded = JWT.decode(token, secret, true, algorithm: ALGORITHM)
    decoded.first.transform_keys(&:to_sym)
  rescue JWT::DecodeError, JWT::ExpiredSignature, JWT::VerificationError
    nil
  end

  def self.secret
    ENV.fetch('JWT_SECRET') do
      raise 'JWT_SECRET environment variable is not set'
    end
  end
  private_class_method :secret
end
